numpy/core/src/multiarray/ctors.c

```diff
@@ -3036,11 +3051,19 @@ static npy_intp
 _calc_length(PyObject *start, PyObject *stop, PyObject *step, PyObject **next, int cmplx)
 {
     npy_intp len, tmp;
-    PyObject *val;
+    PyObject *zero, *val;
+    int next_is_nonzero, val_is_zero;
     double value;
 
+    zero = PyInt_FromLong(0);
+    if (!zero) {
+        return -1;
+    }
+
     *next = PyNumber_Subtract(stop, start);
     if (!(*next)) {
+        Py_DECREF(zero);
```

Comment:
If you move the allocation of `zero` down below this `if`, then you don't need to clear it up here

Reply:
@eric-wieser If I move down `zero` allocation, we need to `DECREF` `*next` and set `NULL` to `next` when the `zero` allocation failed. IMO, this is complicated.

Reply:
Fair

Reply:
>  and set NULL to *next when the zero allocation failed. 

You already need to do this